pkg/virt-launcher/virtwrap/postcopy.go

```diff
@@ -0,0 +1,51 @@
+package virtwrap
+
+import (
+	v1 "kubevirt.io/client-go/api/v1"
+	"kubevirt.io/client-go/log"
+	cmdclient "kubevirt.io/kubevirt/pkg/virt-handler/cmd-client"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/cli"
+
+	libvirt "libvirt.org/libvirt-go"
+)
+
+func (l *LibvirtDomainManager) registerIterationEventForPostCopy(vmi *v1.VirtualMachineInstance, dom cli.VirDomain, options *cmdclient.MigrationOptions, migrationError chan error) {
+	l.virConn.DomainEventMigrationIterationRegister(func(c *libvirt.Connect, d *libvirt.Domain, event *libvirt.DomainEventMigrationIteration) {
+		if event.Iteration == 1 {
+			log.Log.Object(vmi).Info("Signaled start for post copy migration")
+
+			err := dom.MigrateStartPostCopy(uint32(0))
```

Comment:
@ashleyschuett could you please remind me why was this needed? Also, do we have a test for this scenario?

Reply:
All migrations must be started in PreCopy mode, once the migration is started it then needs to be switched to PostCopy.

Reply:
yes... I'm just not so much familial with `.Iteration`. Are we waiting for migration start?
Why is this has to equal to `1`? What happens if we will miss one iteration?

I'm in general worried about libvirt events callbacks. if we are waiting for migration start, we could simply switch to post-copy in the monitoring goroutine.[1]
`libvirt.DOMAIN_JOB_UNBOUNDED` tells us that migration is running and we can switch here.

The advantage is that we actively monitor the sate of the migration in this thread and hopefully will not miss migration state changes.
Also, this is the place where we make all the migration decisions.

What do you think?


[1] https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-launcher/virtwrap/manager.go#L641

Reply:
I choose to watch for the iteration and switch after one iteration since that is how `virsh` implements it when `--postcopy-after-precopy` flag is specified for migration. But it does make sense to have this be more robust in case of an event not being sent from libvirt. 

If we move this to be handled in `liveMigrationMonitor` should we switch to postcopy as soon as an first `libvirt.DOMAIN_JOB_UNBOUNDED` is detected? Or should we still register callback for `DomainEventMigrationIterationRegister` in the  `liveMigraitonMonitor` and then switch on the first iteration or greater if the migration Mode is Pre when it should be Post?

Reply:
> If we move this to be handled in `liveMigrationMonitor` should we switch to postcopy as soon as an first `libvirt.DOMAIN_JOB_UNBOUNDED` is detected?

iirc, `iteration` is iteration over RAM, after which all unlocked pages will be transferred. Without registering a callback, in our case, it would mean that we have observed progress (first time when we update `lastProgressUpdate`)

> Or should we still register callback for `DomainEventMigrationIterationRegister` in the `liveMigraitonMonitor` and then switch on the first iteration or greater if the migration Mode is Pre when it should be Post?

I don't think we should do both.

Moreover, I'm not confident that we want an immediate post-copy migration switch (unless there is a requirement to do so, that I'm not aware of)
Considering all the potential issues that may be associate with our implementation and post-copy migration  - I think the safest would be to switch to post-copy if we didn't manage to migrate in a reasonable amount of time, basically when `acceptableCompletionTime` is reached.
OpenStack has a similar logic [1]

[1] https://github.com/openstack/nova/blob/master/nova/virt/libvirt/driver.py#L9344

Reply:
I am updating this to work as described above. However, my concern is now with the way the mode is specified and will be shown to the user in the VMI status spec object and the UI. 

Right now the configuration key is called `MigrationMode` which can be specified to `PostCopy`. However the migration could complete in `PreCopy` mode if the `acceptableCompletionTime` is never reached and therefore the migration status object of the VMI will then show that the migration is completed and that the mode was `PreCopy` which could be confusing for users. 

I think because of this we should change the name of the configuration key. Maybe `PostCopyEnabled` or if we don't want to be so specific `MigrationModesAllowed` might be better names?  What do you think? Or do you think the naming already makes sense and users will understand?

Reply:
That's a good point. I might be biased at this point, but I don't find this very confusing. But you're right a boolean may be clearer.
I like the `PostCopyEnabled`, but it may be confusing too, as it's not clear whether it is enabled right now or is it enabled in general. How about `PostCopyActive`? (I'm not against PostCopyEnabled, I'll be +1 if you'll choose this)

I went over the code again and I only have one doubt about the configuration field `spec.configuration.migrations.migrationMode: "PostCopy" itself.
Maybe instead of `migrationMode: "PostCopy" ` we should also switch to a boolean, `allowPostCopy` similar to `AllowAutoConverge`, as we don't force it now.
What do you think?
`

Reply:
> I went over the code again and I only have one doubt about the configuration field spec.configuration.migrations.migrationMode: PostCopy itself.
Maybe instead of migrationMode: PostCopy we should also switch to a boolean, allowPostCopy similar to AllowAutoConverge, as we don't force it now.
What do you think? 

Yes this is what i was trying to refer to, but did a poor job explaining. I think `allowPostCopy` is much more clear and will give a better UX experience.